### PR TITLE
PARQUET-1706: [C++] Wrong dictionary_page_offset when writing only data pages via BufferedPageWriter

### PR DESCRIPTION
**Pull request by orozery**
See: https://issues.apache.org/jira/browse/PARQUET-1706

### CODE REVIEW

File: cpp/src/parquet/column_writer.cc [hunk 1]
```diff
@@ -410,25 +410,28 @@ class BufferedPageWriter : public PageWriter {
...
   void Close(bool has_dictionary, bool fallback) override {
     // index_page_offset = -1 since they are not supported
     int64_t final_position = -1;
     PARQUET_THROW_NOT_OK(final_sink_->Tell(&final_position));
-    metadata_->Finish(
-        pager_->num_values(), pager_->dictionary_page_offset() + final_position, -1,
-        pager_->data_page_offset() + final_position, pager_->total_compressed_size(),
-        pager_->total_uncompressed_size(), has_dictionary, fallback);
+    auto dictionary_page_offset =
+        has_dictionary_pages ? pager_->dictionary_page_offset() + final_position : 0;
```
Comment (wesm) [inline cpp/src/parquet/column_writer.cc]: It's a little bit janky that 0 is used as a "do not set the field" sentinel to the `Finish` method. Probably should add a comment here at least about this

File: cpp/src/parquet/column_writer.cc [hunk 2]
```diff
@@ -454,6 +457,7 @@ class BufferedPageWriter : public PageWriter {
   ColumnChunkMetaDataBuilder* metadata_;
   std::shared_ptr<arrow::io::BufferOutputStream> in_memory_sink_;
   std::unique_ptr<SerializedPageWriter> pager_;
+  bool has_dictionary_pages;
```
Comment (wesm) [inline cpp/src/parquet/column_writer.cc]: add trailing `_`

**Review by wesm — COMMENTED** [review]
Thanks for fixing this. I think there have been a couple other patches around this issue over the last couple years but which never had unit tests

**Comment by orozery** [comment]
@ursabot crossbow submit test-conda-cpp

**Review by wesm — APPROVED** [review]
+1, thanks for fixing this

**Pull request closed without merging**